opm/simulators/linalg/getTrueImpesWeights.hpp

```diff
@@ -0,0 +1,131 @@
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with OPM.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#ifndef OPM_GET_QUASI_IMPES_WEIGHTS_HEADER_INCLUDED
+#define OPM_GET_QUASI_IMPES_WEIGHTS_HEADER_INCLUDED
+
+#include <algorithm>
+#include <cmath>
+
+namespace Opm
+{
+
+namespace Details
+{
+    template <class DenseMatrix>
+    DenseMatrix transposeDenseMatrix(const DenseMatrix& M)
+    {
+        DenseMatrix tmp;
+        for (int i = 0; i < M.rows; ++i)
+            for (int j = 0; j < M.cols; ++j)
+                tmp[j][i] = M[i][j];
+
+        return tmp;
+    }
+} // namespace Details
+
+namespace Amg
+{
+            Vector getStorageWeights() const
+        {
+            Vector weights(rhs_->size());
+            BlockVector rhs(0.0);
+            rhs[pressureVarIndex] = 1.0;
+            int index = 0;
+            ElementContext elemCtx(simulator_);
+            const auto& vanguard = simulator_.vanguard();
+            auto elemIt = vanguard.gridView().template begin</*codim=*/0>();
+            const auto& elemEndIt = vanguard.gridView().template end</*codim=*/0>();
+            for (; elemIt != elemEndIt; ++elemIt) {
+                const Element& elem = *elemIt;
+                elemCtx.updatePrimaryStencil(elem);
+                elemCtx.updatePrimaryIntensiveQuantities(/*timeIdx=*/0);
+                Dune::FieldVector<Evaluation, numEq> storage;
+                unsigned threadId = ThreadManager::threadId();
+                simulator_.model().localLinearizer(threadId).localResidual().computeStorage(storage,elemCtx,/*spaceIdx=*/0, /*timeIdx=*/0);
+                Scalar extrusionFactor = elemCtx.intensiveQuantities(0, /*timeIdx=*/0).extrusionFactor();
+                Scalar scvVolume = elemCtx.stencil(/*timeIdx=*/0).subControlVolume(0).volume() * extrusionFactor;
+                Scalar storage_scale = scvVolume / elemCtx.simulator().timeStepSize();
+                MatrixBlockType block;
+                double pressure_scale = 50e5;
+                for (int ii = 0; ii < numEq; ++ii) {
+                    for (int jj = 0; jj < numEq; ++jj) {
+                        block[ii][jj] = storage[ii].derivative(jj)/storage_scale;
+                        if (jj == pressureVarIndex) {
+                            block[ii][jj] *= pressure_scale;
+                        }
+                    }
+                }
+                BlockVector bweights;
+                MatrixBlockType block_transpose = Opm::transposeDenseMatrix(block);
+                block_transpose.solve(bweights, rhs);
+                bweights /= 1000.0; // given normal densities this scales weights to about 1.
+                weights[index] = bweights;
+                ++index;
+            }
+            return weights;
```

Comment:
Please double check that the computation in the above loop does not depend on values of the cell neighors. If it does the we might need some communication before returning the weights.

Reply:
As long the the diagonal of the matrix is calculated correctly for the owning rows I think it is correct.